netty/src/main/java/io/grpc/netty/GrpcHttp2ConnectionHandler.java

```diff
@@ -69,6 +74,26 @@ public void handleProtocolNegotiationCompleted(
       Attributes attrs, InternalChannelz.Security securityInfo) {
   }
 
+  /**
+   * Returns the channel logger for the given channel context, or a Noop Logger if absent.
+   */
+  public ChannelLogger getNegotiationLogger() {
+    if (negotiationLogger != null) {
+      return negotiationLogger;
+    }
+    // This is only for tests where there may not be a valid logger.
```

Comment:
Would it be too hard to have tests pass this in themselves? We can add this implementation to core/src/test/io/grpc/internal for all tests to use.

Reply:
It is not hard. But I was thinking that this default also provides the benefits of avoiding NPE when it is null, like in NettyServerHandle case. There is no ChannelLogger at server side, and if the get method is accidentally called it would cause NPE.

Reply:
Oh, I see... Hmm... It seems if there is any caller of this on server-side we _should_ throw an exception, because it doesn't work like the code wants. Don't return null; actually throw IllegalStateException or something like that. I'd agree the NPE doesn't sound great, but I think the solution is make it a _bigger_ error, instead of trying cover over it. But I could live with this.

Reply:
Interestingly we do have many usages(*illegal*) of the server side logger, and they fallback to the default noop . It is hard to prevent them from calling the *get* method, because negotiation handlers are often shared between client and server. 
So throwing errors immediately at *get* might be too strict for now.
So first option we return null, and check null when they need to use the logger, verbose but flexible.  
Alternatively, it looks to me that the current way of providing default at *get* method is not too bad. Wdyt?

Reply:
Looking more deeply at the change, I see how previously all usages used a helper to be provided a non-null instance. So this is fairly 1:1 with the existing behavior. But the existing behavior is sort of botched and strongly misleading.

Many of the negotiators are client-only or server-only, so it is strange to me that we'd have much confusion. It looks like ProtocolNegotiationHandler, WaitUntilActiveHandler, and TsiHandshakerHandler are the places it is confused. ProtocolNegotiationHandler is used in many handshakers internally, so it will have annoying consequences.

I think we actually want to log on server-side in these places. And it would be annoying to have a separate log code path at each site that needs to log. So although it is weird to provide a "channel logger" on server-side, I think that may be the best path forward. Instead of having a noop logger it will just use normal java.util.logging, using [the same level conversion](https://github.com/grpc/grpc-java/blob/e5ab4d743da0e9268905369a83f2c4faeee77d4e/core/src/main/java/io/grpc/internal/ChannelLoggerImpl.java#L97) as on client-side.

It looks like this change will require fixups to a reasonable amount of code internally. I'd encourage you to have those changes ready and merge the internal fixes just after this is merged. (Although you may delay on making those changes internally after this reviews of this PR look like they are settling down.)